gnu_dataset/gnats-4.2.0/doc/man/pr-edit.man
.\" Copyright (c) 2000, 2003 Free Software Foundation, Inc.
.\" See section COPYING for conditions for redistribution
.TH pr-edit 8 "August 2003" "GNATS 4.2.0" "GNATS Admininstration Utilities"
.de BP
.sp
.ti -.2i
\(**
..
.SH NAME
pr\-edit \- creates, edits or deletes PRs
.SH SYNOPSIS
.hy 0
.na
.TP
.B pr\-edit
[\ \fB\-\-lock\fR\=\fIusername\fR\ |\ \fB\-l\fR\ \fIusername\fR\ ]
[\ \fB\-\-unlock\fR\ |\ \fB\-u\fR\ ]
.br
[\ \fB\-\-lockdb\fR\ |\ \fB\-L\fR\ ]\ [\ \fB\-\-unlockdb\fR\ |\ \fB\-U\fR\ ]
.br
[\ \fB\-\-check\fR\ |\ \fB\-c\fR\ ]
[\ \fB\-\-check\-initial\fR\ |\ \fB\-C\fR\ ]
.br
[\ \fB\-\-submit\fR\ |\ \fB\-s\fR\ ]
.br
[\ \fB\-\-append\fR\=\fIfield\fR\ |\ \fB\-a\fR\ \fIfield\fR\ ]
[\ \fB\-\-replace\fR\=\fIfield\fR\ |\ \fB\-r\fR\ \fIfield\fR\ ]
.br
[\ \fB\-\-reason\fR\=\fIreason\fR\ |\ \fB\-R\fR\ \fIreason\fR\ ]
.br
[\ \fB\-\-delete\-pr\fR\ ]
.br
[\ \fB\-\-process\fR\=\fIprocess\-id\fR\ |\ \fB\-p\fR\ \fIprocess\-id\fR\ ]
.br
[\ \fB\-\-database\fR\=\fIdatabase\fR\ |\ \fB\-d\fR\ \fIdatabase\fR\ ]
.br
[\ \fB\-\-filename\fR\=\fIfilename\fR\ |\ \fB\-f\fR\ \fIfilename\fR\ ]
.br
[\ \fB\-\-version\fR\ |\ \fB\-V\fR\ ]
[\ \fB\-\-help\fR\ |\ \fB\-h\fR\ ]
.br
[\ \fB\-\-user\fR\=\fIusername\fR\ |\ \fB\-v\fR\ \fIusername\fR\ ]
.br
[\ \fB\-\-passwd\fR\=\fIpassword\fR\ |\ \fB\-w\fR\ \fIpassword\fR\ ]
.br
[\ \fB\-\-host\fR\=\fIhost\fR\ |\ \fB\-H\fR\ \fIhost\fR\ ]
.br
[\ \fB\-\-port\fR\=\fIport\fR\ |\ \fB\-P\fR\ \fIport\fR\ ]
.br
[\ \fB\-\-debug\fR\ |\ \fB\-D\fR\ ]
.br
[\ \fIPR\fR\ ]
.ad b
.hy 1
.SH DESCRIPTION
\fBpr\-edit\fR can be used to edit an existing \fIPR\fR by either
replacing or appending to particular fields within the \fIPR\fR, or
providing a new \fIPR\fR to replace the existing one. \fBpr\-edit\fR
can also be used to create new \fIPR\fRs from scratch, or delete
existing ones.
.P
\fBpr\-edit\fR also provides miscellaneous services for locking and
unlocking \fIPR\fRs,  locking or unlocking an entire database, or
verifying that proposed \fIPR\fR contents are valid and correct.
.SH OPTIONS
.TP 0.5i
\fB\-\-lockdb\fR,\ \fB\-L\fR
Locks the specified database.  No PRs may be edited, created or deleted
while the database is locked.
.LP
.RS 0.5i
This option is generally used when editing the index file.
.RE
.TP 0.5i
\fB\-\-unlockdb\fR,\ \fB\-U\fR
Unlocks the database.  No check is made that the invoking user actually had
locked the database in the first place; hence, it is possible for anyone
to steal a database lock.
.TP
\fB\-\-check\fR,\ \fB\-c\fR
.TP
\fB\-\-check\-initial\fR,\ \fB\-C\fR
The \fB\-\-check\fR options are used to verify that a proposed
\fIPR\fR's field contents are valid.  The \fIPR\fR is read in (either
from stdin or a file specified with \fB\-\-filename\fR), and its
fields are compared against the rules specified by the database
configuration of the selected database.  Warnings are given for
enumerated fields whose contents do not contain one of the required
values or fields that do not match required regexps.
.P
.RS 0.5i
\fB\-\-check\-initial\fR is used to verify initial \fIPR\fRs, rather
than proposed edits of existing \fIPRs\fR.
.RE
.TP 0.5i
\fB\-\-submit\fR,\ \fB\-s\fR
Used to submit a new \fIPR\fR to the database.  The \fIPR\fR is read
in and verified for content; if the \fIPR\fR is valid as an initial
\fIPR\fR, it is then added to the database.
.P
.RS 0.5i
A zero exit code is
returned if the submission was successful.  Otherwise, the reason(s)
for the \fIPR\fR being rejected are printed to stdout, and a non-zero
exit code is returned.
.RE
.P
The following options require a \fIPR number\fR to be given.
.TP 0.5i
\fB\-\-delete\-pr\fR
Deletes the specified \fIPR\fR from the database.  The \fIPR\fR must be
in a closed state, and not locked.
.LP
.RS 0.5i
Only the GNATS user (by default \fBgnats\fR) is permitted to delete \fIPR\fRs.
.RE
.TP 0.5i
\fB\-\-lock\fR\=\fIusername\fR,\ \fB\-l\fR\ \fIusername\fR
Locks the \fIPR\fR.  \fIusername\fR is associated with the lock, so
the system administrator can determine who actually placed the lock on
the \fIPR\fR.  However, anyone is permitted to remove locks on a
\fIPR\fR.
.P
.RS 0.5i
If the optional \fB\-\-process\-id\fR option is also given, that
\fIprocess\-id\fR is associated with the lock.
.RE
.TP 0.5i
\fB\-\-unlock\fR,\ \fB\-u\fR
Unlocks the \fIPR\fR.
.TP
\fB\-\-append\fR\=\fIfield\fR,\ \fB\-a\fR\ \fIfield\fR
.TP
\fB\-\-replace\fR\=\fIfield\fR,\ \fB\-r\fR\ \fIfield\fR
\fB\-\-append\fR and \fB\-\-replace\fR are used to append or replace
content of a specific field within a \fIPR\fR.  The new field content
is read in from stdin (or from the file specified with the
\fB\-\-filename\fR option), and either appended or replaced to the
specified field.  The field contents are verified for correctness
before the \fIPR\fR is rewritten.
.P
.RS 0.5i
If the edit is successful, a zero exit status is returned.  If the edit
failed, a non-zero exit status is returned, and the reasons for the failure
are printed to stdout.
.RE
.TP 0.5i
\fB\fB\-\-reason\fR\=\fIreason\fR,\ \fB\-R\fR\ \fIreason\fR
Certain PR fields are configured in the database configuration to
require a short text describing the reason for every change that is
made to them.  If you edit a PR and change any such fields, you must
provide a short text, the \fIreason\fR for the change, through this
option.  If the option is used and no change-reason requiring field is
actually changed, the option has no effect.
.RE
.TP 0.5i
\fI[PR]\fR
If only a \fIPR\fR number is specified with no other options, a
replacement \fIPR\fR is read in (either from stdin or the file
specified with \fB\-\-filename\fR).  If the \fIPR\fR contents are
valid and correct, the existing \fIPR\fR is replaced with the new
\fIPR\fR contents.
.P
.RS 0.5i
If the edit is successful, a zero exit status is returned.  If the edit
failed, a non-zero exit status is returned, and the reasons for the failure
are printed to stdout.
.RE
.TP 0.5i
\fB\-\-database\fR\=\fIdatabase\fR,\ \fB\-d\fR\ \fIdatabase\fR
Specifies the database which is to be manipulated.  If no database is
specified, the database named \fIdefault\fR is assumed.  This option
overrides the database specified in the \fBGNATSDB\fR environment
variable.
.TP 0.5i
\fB\-\-filename\fR\=\fIfilename\fR,\ \fB\-f\fR\ \fIfilename\fR
For actions that require reading in a \fIPR\fR or field content, this
specifies the name of a file to read.  If \fB\-\-filename\fR is not
specified, the \fIPR\fR or field content is read in from stdin.
.TP 0.5i
\fB\-\-version\fR,\ \fB\-V\fR
Displays the version number of the program.
.TP 0.5i
\fB\-\-help\fR,\ \fB\-h\fR
Prints a brief usage message.
.TP 0.5i
\fB\-\-host\fR\=\fIhost\fR,\ \fB\-H\fR\ \fIhost\fR
Hostname of the GNATS server.
.TP 0.5i
\fB\-\-port\fR\=\fIport\fR,\ \fB\-P\fR\ \fIport\fR
The port that the GNATS server runs on.
.TP 0.5i
\fB\-\-user\fR\=\fIusername\fR,\ \fB\-v\fR\ \fIusername\fR
Username used when logging into the GNATS server.
.TP 0.5i
\fB\-\-passwd\fR\=\fIpassword\fR,\ \fB\-w\fR\ \fIpassword\fR
Password used when logging into the GNATS server.
.TP 0.5i
\fB\-\-debug\fR,\ \fB\-D\fR
Used to debug network connections.
.SH ENVIRONMENT VARIABLES
The \fBGNATSDB\fR environment variable is used to determine which database to
use.  For a local database, it contains the name of the database to access.
.P
For network access via \fIgnatsd\fR, it contains a colon-separated list
of strings that describe the remote database, in the form
.TP
\fIserver\fR:\fIport\fR:\fIdatabasename\fR:\fIusername\fR:\fIpassword\fR
.P
Any of the fields may be omitted, but at least one colon must appear;
otherwise, the value is assumed to be the name of a local database.
.P
If \fBGNATSDB\fR is not set and the --database option is not supplied,
it is assumed that the database is local and that its name is
\fIdefault\fR.
.SH "SEE ALSO"
.I Keeping Track: Managing Messages With GNATS
(also installed as the GNU Info file
.BR gnats.info )
.LP
.BR databases (5),
.BR dbconfig (5),
.BR delete-pr (8),
.BR edit-pr (1)
.BR file-pr (8),
.BR gen-index (8),
.BR gnats (7),
.BR gnatsd (8),
.BR mkcat (8),
.BR mkdb (8),
.BR pr-edit (8),
.BR query-pr (1),
.BR queue-pr (8),
.BR send-pr (1).
.SH COPYING
Copyright (c) 2000, 2003, Free Software Foundation, Inc.
.PP
Permission is granted to make and distribute verbatim copies of
this manual provided the copyright notice and this permission notice
are preserved on all copies.
.PP
Permission is granted to copy and distribute modified versions of this
manual under the conditions for verbatim copying, provided that the
entire resulting derived work is distributed under the terms of a
permission notice identical to this one.
.PP
Permission is granted to copy and distribute translations of this
manual into another language, under the above conditions for modified
versions, except that this permission notice may be included in
translations approved by the Free Software Foundation instead of in
the original English.


gnu_dataset/gnats-4.2.0/gnats/pr-edit.c
/* Entry for the GNATS system.
   Copyright (C) 2001 Milan Zamazal
   Copyright (C) 1993, 1994, 1995, 2007 Free Software Foundation, Inc.
   Originally written by Tim Wicinski (wicinski@barn.com).
   Further work by Brendan Kehoe (brendan@cygnus.com).
   Further work by Milan Zamazal (pdm@zamazal.org).

This file is part of GNU GNATS.

GNU GNATS is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 3, or (at your option)
any later version.

GNU GNATS is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with GNU GNATS; see the file COPYING. If not, see
<http://www.gnu.org/licenses/>.
*/

#include "gnats.h"
#include "pcodes.h"
#include "ds.h"

/* Name of this program.  */
const char *program_name;

enum {
  NONE,
  LOCK,
  UNLOCK,
  LOCKDB,
  UNLOCKDB,
  CHECK,
  CHECK_INITIAL,
  MODIFY,
  SUBMIT,
  APPEND,
  REPLACE,
  DELETE
} edit_options;

#define DELETE_PR_OPT 256
#define SHOW_PRNUM_OPT 257

struct option long_options[] =
{
  {"lock", 1, NULL, 'l'},
  {"unlock", 0, NULL, 'u'},
  {"lockdb", 0, NULL, 'L'},
  {"unlockdb", 0, NULL, 'U'},
  {"check", 0, NULL, 'c'},
  {"check-initial", 0, NULL, 'C'},
  {"submit", 0, NULL, 's'},
  {"append", 1, NULL, 'a'},
  {"replace", 1, NULL, 'r'},
  {"reason", 1, NULL, 'R'},
  {"process", 1, NULL, 'p'},
  {"database", 1, NULL, 'd'},
  {"filename", 1, NULL, 'f'},
  {"version", 0, NULL, 'V'},
  {"delete-pr", 0, NULL, DELETE_PR_OPT},
  {"show-prnum", 0, NULL, SHOW_PRNUM_OPT},
  {"help", 0, NULL, 'h'},
  {"user", 1, NULL, 'v'},
  {"passwd", 1, NULL, 'w'},
  {"host", 1, NULL, 'H'},
  {"port", 1, NULL, 'P'},
  {"debug", 0, NULL, 'D'},
  {"email-addr", 1, NULL, 'e'},
  {"global-databases-file",1,NULL,'g'},
  {NULL, 0, NULL, 0}
};

#define PROGRAM_NAME "pr-edit"

static const char *const USAGE[] = {
  "Usage: " PROGRAM_NAME " [OPTION]... [PR]\n\
Modify database.\n\
\n",
"Actions:\n\
  -l --lock=USER           lock the given PR\n\
  -u --unlock              unlock the given PR\n\
  -L --lockdb              lock the whole database\n\
  -U --unlockdb            unlock the database\n\
  -c --check               check input for editting, don't change PR\n\
  -C --check-initial       check input for submission, don't submit new PR\n\
  -s --submit              submit new PR\n",
  "\
     --show-prnum          display the newly created PR number (for --submit)\n\
  -a --append=FIELDNAME    append input to FIELDNAME\n\
  -r --replace=FIELDNAME   replace FIELDNAME with input\n\
     --delete-pr           delete PR from the database completely\n\
  -R --reason=REASON       change describing text required by some fields\n\
\n",
"Connection options:\n\
  -d --database=DATABASE   use DATABASE\n\
  -H --host=SERVER         connect to SERVER\n\
  -P --port=PORT           connect to the port PORT\n\
  -v --user=NETID          use NETID as the username on the remote server\n\
  -w --passwd=PASSWORD     user's password\n\
  -e --email-addr=ADDRESS  send contingent e-mail reports to ADDRESS\n\
  -p --process=PID         PID of the connecting process\n\
  -f --filename=FILENAME   read input from FILENAME instead of stdin\n\
\n",
"Other options:\n\
  -h --help                display this help and exit\n\
  -V --version             output version information and exit\n\
  -D --debug               enable some debugging\n\
  -g --global-databases-file=PATH\n\
                           PATH to databases file\n",
  NULL
};   


/* Read the entire contents of the file referenced by INP into memory, and
   return it as a char* string allocated with malloc ().  */

static char *
read_file (FILE *inp)
{
  char *res = xmalloc (512);  /* 512 is just a number.  It's not a limit of
				 any sort.  */
  size_t reslen = 512;
  size_t totlen = 0, inplen;

  while ((inplen = fread (res + totlen, 1, reslen - totlen, inp)) != 0)
    {
      totlen += inplen;
      if (totlen == reslen)
	{
	  reslen += 512;
	  res = xrealloc (res, reslen);
	}
    }
  if (totlen == reslen)
    {
      reslen++;
      res = xrealloc (res, reslen);
    }
  if (inp != stdin)
    {
      fclose (inp);
    }
  res[totlen] = '\0';
  return res;
}

static void
handleNetworkEdit (int edit_options, FILE *fpin, char *prnum, char *username,
		   char *editEmailAddr, char *processid, char *fieldname,
		   char *reason, int show_prnum)
{
  int exitcode = 0;

  switch (edit_options)
    {
    case LOCK:
      netLockPR (prnum, username, processid);
      break;
    case UNLOCK:
      netUnlockPR (prnum);
      break;
    case APPEND:
      netEditField (fpin, prnum, fieldname, editEmailAddr, 1, reason);
      break;
    case REPLACE:
      netEditField (fpin, prnum, fieldname, editEmailAddr, 0, reason);
      break;
    case SUBMIT:
      {
	netSubmitNewPR (fpin, show_prnum);
	break;
      }
    case CHECK:
    case CHECK_INITIAL:
      {
	netCheckPR (fpin, edit_options == CHECK_INITIAL ? 1 : 0);
	break;
      }
    case LOCKDB:
      netLockDB ();
      break;
    case UNLOCKDB:
      netUnlockDB ();
      break;
    case DELETE:
      netDeletePR (prnum, editEmailAddr);
      break;
    default:
      {
	netModifyPR (fpin, prnum, editEmailAddr);
	break;
      }
    }
  client_exit ();
  exit (exitcode);
}

extern int debug;

int
main (int argc, char **argv)
{
  DatabaseInfo database = NULL;
  FILE *fp = stdin;
  int option;
  char *prnum = NULL;
  char *username = (char *)NULL, *processid = (char *)NULL;
  /* Default to 1, a passing value.  */
  int result = 1;
  ErrorDesc err;
  int db_locked = 0;
  char *fieldname = NULL;
  char *nameOfDatabase = NULL;
  char *netid = NULL;
  char *passwd = NULL;
  char *host = NULL;
  int port = -1;
  int networkmode = 0;
  char *editUserEmailAddr = NULL;
  char *reason = NULL;
  int show_prnum = 0;

  program_name = basename (argv[0]);
  edit_options = MODIFY;

  while ((option = getopt_long (argc, argv, "l:uLUDhcp:d:g:f:VFv:w:H:P:e:",
				long_options, (int *)0)) != EOF)
    {
      switch (option)
	{
	case 'd':
	  nameOfDatabase = optarg;
	  break;

	case 'g':
	  if (optarg[0] == '\0')
	    {
              fprintf(stderr, "%s: filename must be non-null\n", program_name);
              exit(1);
	    }
	  global_db_list_file = optarg;
	  break;

	case 'D':
	  debug = 1;
	  break;

	case SHOW_PRNUM_OPT:
	  show_prnum = 1;
	  break;

	case 'f':
	  fp = fopen (optarg, "r");
	  /* If they gave a bogus argument, then exit right away; we don't
	     want to send mail to gnats-admin every time someone tries to
	     edit a bogus PR number.  */
	  if (fp == (FILE *)NULL)
            {
              fprintf (stderr, "%s: Bad PR file %s.\n", program_name, optarg);
              exit (2);
            }
	  break;

	case 'p':
	  processid = optarg;
	  break;

	case 'l': 
	  edit_options = LOCK;
	  username = optarg;
	  break;

	case 'u':
	  edit_options = UNLOCK;
	  break;

	case 'L':
	  edit_options = LOCKDB;
	  break;

	case 'U':
	  edit_options = UNLOCKDB;
	  break;

	case 'c':
	  edit_options = CHECK;
	  break;

	case 'C':
	  edit_options = CHECK_INITIAL;
	  break;

	case 's':
	  edit_options = SUBMIT;
	  break;

	case 'a':
	  edit_options = APPEND;
	  fieldname = optarg;
	  break;

	case 'r':
	  edit_options = REPLACE;
	  fieldname = optarg;
	  break;

	case 'R':
	  reason = optarg;
	  break;
	  
	case DELETE_PR_OPT:
	  edit_options = DELETE;
	  break;

	case 'V':
	  version (PROGRAM_NAME);
	  exit (0);
	  break;
	  
	case 'h':
	  usage (USAGE, 0);
	  break;

	case 'H':
	  host = optarg;
	  networkmode = 1;
	  break;

	case 'P':
	  port = atoi (optarg);
	  networkmode = 1;
	  break;

        case 'v':
          netid = optarg;
	  networkmode = 1;
          break;

        case 'w':
          passwd = optarg;
	  networkmode = 1;
          break;

	case 'e':
	  editUserEmailAddr = xstrdup (optarg);
	  break;

	default:
	  usage (USAGE, 3);
	  break;
	}
    }

  /* if there is another arg, then we take it to be a file name. */
  if (optind < argc)
    {
      prnum = argv[optind];
    }

  if (prnum == NULL
      && (edit_options == MODIFY
	  || edit_options == APPEND
	  || edit_options == REPLACE))
    {
      fprintf (stderr, "%s: Must supply the PR number when modifying a PR\n",
	       program_name);
      exit (1);
    }

  if (editUserEmailAddr == NULL)
    {
      struct passwd *myent = getpwuid (getuid ());
      if (myent != NULL)
	{
	  char hostname[256];

	  gethostname (hostname, sizeof (hostname));
	  asprintf (&editUserEmailAddr, "%s@%s", myent->pw_name, hostname);
	}
      else
	{
	  fprintf (stderr, "%s: no username for uid %d\n", program_name,
		   (int) getuid ());
	  exit (1);
	}
    }

  if ((edit_options == LOCK) && (prnum == NULL || ! username))
    {
      fprintf (stderr,
	       "%s: Must lock with PR number, username, and process locking it.\n",
               program_name);
      exit (2);
    }

  if ((edit_options == UNLOCK) && (prnum == NULL))
    {
      fprintf (stderr, "%s: Must unlock with PR number.\n",
               program_name);
      exit (2);
    }

  if (! networkmode)
    {
      networkmode = gnatsdbHasNetconn (nameOfDatabase);
    }

  if (networkmode)
    {
      if (client_init_gnats (&err, netid, passwd, host, port,
			     nameOfDatabase) != 0)
	{
	  client_print_errors (database, err);
	  exit (3);
	}
      handleNetworkEdit (edit_options, fp, prnum, username, editUserEmailAddr,
			 processid, fieldname, reason, show_prnum);
    }

  database = init_gnats (program_name, nameOfDatabase, &err);
  if (database == NULL)
    {
      result = 0;
    }
  else
    {
      umask (022);

      block_signals ();

      if (edit_options == LOCKDB
	  || edit_options == APPEND
	  || edit_options == REPLACE
	  || edit_options == MODIFY
	  || edit_options == SUBMIT)
	{
	  result = (client_lock_gnats (database, &err) == 0);
	  if (result)
	    {
	      db_locked = 1;
	    }
	}
    }
  
  if (result)
    {
      switch (edit_options) 
	{
	case LOCK:
	  if (pr_exists (database, prnum, &err))
	    {
	      result = lock_pr (database, prnum, username, processid, &err);
	      if (result != 0)
		{
		  PR *pr = pr_load_by_id (database, prnum, 0, &err);
		  if (pr != NULL)
		    {
		      print_named_format_pr (stdout, pr, "edit", "\n", &err);
		      free_pr (pr);
		    }
		  else
		    {
		      if (unlock_pr (database, prnum, &err) != 0)
			{
			  /* ??? XXX !!! */
			  setError (&err, CODE_NONEXISTENT_PR,
				    "Invalid PR %s.", prnum);
			}
		      result = 0;
		    }
		}
	    }
	  else
	    {
	      result = 0;
	    }
	  break;

	case UNLOCK:
	  result = unlock_pr (database, prnum, &err);
	  break;

	case SUBMIT:
	  {
	    result = submit_pr (database, fp, &err);
            if (show_prnum && result)
              {
                fprintf (stdout, "%d\n", result);
              }
	    break;
	  }
	case APPEND:
	  {
	    char *contents = read_file (fp);
	    FieldIndex field = find_field_index (database, fieldname);
	    if (field == InvalidFieldIndex)
	      {
		setError (&err, CODE_INVALID_FIELD_NAME,
			  "No such field as %s", fieldname);
		result = 0;
	      }
	    else
	      {
		result = edit_field (database, prnum, field, 1,
				     contents, reason,
				     editUserEmailAddr, &err);
	      }
	    break;
	  }
	case REPLACE:
	  {
	    char *contents = read_file (fp);

	    FieldIndex field = find_field_index (database, fieldname);
	    if (field == InvalidFieldIndex)
	      {
		setError (&err, CODE_INVALID_FIELD_NAME,
			  "No such field as %s", fieldname);
		result = 0;
	      }
	    else
	      {
		result = edit_field (database, prnum, field, 0, 
				     contents, reason,
				     editUserEmailAddr, &err);
	      }
	    break;
	  }
	case DELETE:
	  {
	    struct passwd *pwent = getpwnam (GNATS_USER);

	    if (pwent == NULL)
	      {
		setError (&err, CODE_ERROR,
			  "No such user as %s\n", GNATS_USER);
		result = 0;
	      }
	    else if ((int) pwent->pw_uid != (int) getuid ())
	      {
		setError (&err, CODE_FILE_ERROR,
			  "Must be user %s to delete a PR\n", GNATS_USER);
		result = 0;
	      }
	    else
	      {
		result = ! deletePR (database, prnum, editUserEmailAddr, &err);
	      }
	    break;
	  }
	case CHECK:
	case CHECK_INITIAL:
	  result = check_pr_file (database, fp, &err, 
				  edit_options == CHECK_INITIAL);
	  break;
	case LOCKDB:
	case UNLOCKDB:
	  result = 1; /* so that the program exits with 0 status */
	  break;

	default: 
	  result = replace_pr (database, fp, editUserEmailAddr, &err);
	  break;
	}
    }

  if (db_locked || edit_options == UNLOCKDB)
    {
      client_unlock_gnats ();
    }

  if (! result)
    {
      client_print_errors (database, err);
    }

  unblock_signals ();

  if (editUserEmailAddr != NULL)
    {
      free (editUserEmailAddr);
      editUserEmailAddr = NULL;
    }

  exit (result ? 0 : 2);
}
